qa/rpc-tests/bip135-threshold.py

```diff
@@ -0,0 +1,317 @@
+from test_framework.mininode import CTransaction, NetworkThread
+from test_framework.blocktools import create_coinbase, create_block
+from test_framework.comptool import TestInstance, TestManager
+from test_framework.script import CScript, OP_1NEGATE, OP_CHECKSEQUENCEVERIFY, OP_DROP
+from io import BytesIO
+import time
+import itertools
+import tempfile
+
+'''
+This test exercises BIP135 fork activation thresholds.
+It uses a single node with custom forks.csv file.
+
+It is originally derived from bip9-softforks.
+'''
+
+
+class BIP135ForksTest(ComparisonTestFramework):
+
+    def __init__(self):
+        super().__init__()
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.defined_forks = [ "bip135test%d" % i for i in range(0,8) ]
+
+    def setup_network(self):
+        '''
+        sets up with a custom forks.csv to test threshold / grace conditions
+        also recomputes deployment start times since node is shut down and
+        restarted between main test sections.
+        '''
+        # write a custom forks.csv file containing test deployments.
+        # blocks are 1 second apart by default in this regtest
+        fork_csv_filename = os.path.join(self.options.tmpdir, "forks.csv")
+        # forks.csv fields:
+        # network,bit,name,starttime,timeout,windowsize,threshold,minlockedblocks,minlockedtime,gbtforce
+        with open(fork_csv_filename, 'wt') as fh:
+            # use current time to compute offset starttimes
+            self.init_time = int(time.time())
+            # starttimes are offset by 30 seconds from init_time
+            self.fork_starttime = self.init_time + 30
+            fh.write(
+            "# deployment info for network 'regtest':\n" +
+
+            ########## THRESHOLD TESTING BITS (1-6) ############
+
+            # bit 0: test 'csv' fork renaming/reparameterization
+            "regtest,0,bip135test0,%d,999999999999,144,108,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 1: test 'segwit' fork renaming/reparameterization
+            "regtest,1,bip135test1,%d,999999999999,144,108,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 2: test minimum threshold
+            "regtest,2,bip135test2,%d,999999999999,100,1,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 3: small threshold
+            "regtest,3,bip135test3,%d,999999999999,100,10,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 4: supermajority threshold
+            "regtest,4,bip135test4,%d,999999999999,100,75,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 5: high threshold
+            "regtest,5,bip135test5,%d,999999999999,100,95,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 6: max-but-one threshold
+            "regtest,6,bip135test6,%d,999999999999,100,99,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 7: max threshold
+            "regtest,7,bip135test7,%d,999999999999,100,100,0,0,true\n" % (self.fork_starttime)
```

Comment:
I would like to see a test that verifies:
 
1. the timeout.  The test should produce threshold-1 blocks in windowsize before timeout, and then produces a block with the bit set after timeout.  This should NOT activate

2. Previously I mentioned confusion as to when activation bits were counted.  If they are counted only when blockheight%windowsize==0, I would like to see a test that produces more than threshold blocks with the bit set, but does NOT activate because these blocks cross a windowsize boundary.

We need to test these specification edge conditions in the python code because the python code can easily be used for interoperability testing (between full nodes), whereas the C code is quite locked into a specific implementation.